modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITThinClientConnectionTest.java

```diff
@@ -0,0 +1,173 @@
+                "  }\n" +
+                "}");
+        }};
+
+    /** */
+    private final List<Ignite> startedNodes = new ArrayList<>();
+
+    /** */
+    @AfterEach
+    void tearDown() throws Exception {
+        IgniteUtils.closeAll(Lists.reverse(startedNodes));
+    }
+
+    /**
+     * Check that thin client can connect to any server node and work with table API.
+     */
+    @Test
+    void testThinClientConnectsToServerNodesAndExecutesBasicTableOperations() throws Exception {
+        nodesBootstrapCfg.forEach((nodeName, configStr) ->
+                startedNodes.add(IgnitionManager.start(nodeName, configStr, workDir.resolve(nodeName)))
+        );
+
+        var keyCol = "key";
+        var valCol = "val";
+
+        SchemaTable schTbl = SchemaBuilders.tableBuilder(SCHEMA_NAME, TABLE_NAME).columns(
+                SchemaBuilders.column(keyCol, ColumnType.INT32).asNonNull().build(),
+                SchemaBuilders.column(valCol, ColumnType.string()).asNullable().build()
+        ).withPrimaryKey(keyCol).build();
+
+        startedNodes.get(0).tables().createTable(schTbl.canonicalName(), tblCh ->
+                SchemaConfigurationConverter.convert(schTbl, tblCh)
+                        .changeReplicas(1)
+                        .changePartitions(10)
+        );
+
+        var addrs = new String[]{"127.0.0.1:" + getClientPort("node0"), "127.0.0.1:" + getClientPort("node1")};
+
+        for (var addr : addrs) {
+            try (var client = IgniteClient.builder().addresses(addr).build()) {
+                List<Table> tables = client.tables().tables();
+                assertEquals(1, tables.size());
+
+                Table table = tables.get(0);
+                assertEquals(String.format("%s.%s", SCHEMA_NAME, TABLE_NAME), table.tableName());
+
+                var tuple = table.tupleBuilder().set(keyCol, 1).set(valCol, "Hello").build();
+                var keyTuple = table.tupleBuilder().set(keyCol, 1).build();
+
+                table.upsert(tuple);
+                assertEquals("Hello", table.get(keyTuple).stringValue(valCol));
+
+                assertTrue(table.delete(keyTuple));
+            }
+        }
+    }
+
+    /**
+     * Gets the client listener port.
+     *
+     * @param nodeName Node name.
+     * @return Port number.
+     * @throws Exception When failed.
+     */
+    private static int getClientPort(String nodeName) throws Exception {
+        InetSocketAddress addr = (InetSocketAddress) getClientHandlerModule(nodeName).localAddress();
+        assertNotNull(addr);
+
+        return addr.getPort();
+    }
+
+    /**
+     * Gets the client handler module for the give node.
+     *
+     * @param nodeName Node name.
+     * @return Client handler module.
+     * @throws Exception When failed.
+     */
+    private static ClientHandlerModule getClientHandlerModule(String nodeName) throws Exception {
+        Field field = IgnitionImpl.class.getDeclaredField("nodesStartedComponents");
```

Comment:
Why just didn't you set the client port in node config? As I see, other network ports are hardcoded in config already.

Reply:
There is no guarantee that the server will bind to the given port.

I've actually used a hardcoded port initially and then spent some time debugging test hangs on TeamCity, where `10800` turned out to be occupied by something else.

Client handled has `port` + `portRange` and will bind to the first available port. The only way to ensure stable tests is to retrieve the actual bound port from the module.